Test/spec/simd/simd_f32x4.wast

```diff
@@ -2119,6 +2137,36 @@
                                                         (v128.const f32x4 nan:0x200000 nan:0x200000 nan:0x200000 nan:0x200000)))
 (assert_return_arithmetic_nan_f32x4 (invoke "f32x4.max" (v128.const f32x4 -nan:0x200000 -nan:0x200000 -nan:0x200000 -nan:0x200000)
                                                         (v128.const f32x4 -nan:0x200000 -nan:0x200000 -nan:0x200000 -nan:0x200000)))
+
+;; Test different lanes go through different if-then clauses
+(assert_return (invoke "f32x4.min" (v128.const f32x4 nan 0 0 1)
+                                   (v128.const f32x4 0 -nan 1 0))
+                                   (v128.const f32x4 nan -nan 0 0))
+(assert_return (invoke "f32x4.min" (v128.const f32x4 nan 0 0 0)
+                                   (v128.const f32x4 0 -nan 1 0))
+                                   (v128.const f32x4 nan -nan 0 0))
+(assert_return (invoke "f32x4.max" (v128.const f32x4 nan 0 0 1)
+                                   (v128.const f32x4 0 -nan 1 0))
+                                   (v128.const f32x4 nan -nan 1 1))
+(assert_return (invoke "f32x4.max" (v128.const f32x4 nan 0 0 0)
+                                   (v128.const f32x4 0 -nan 1 0))
+                                   (v128.const f32x4 nan -nan 1 0))
+
+;; Test opposite signs of zero
+(assert_return (invoke "f32x4.min" (v128.const f32x4 0 0 -0 +0)
+                                   (v128.const f32x4 +0 -0 +0 -0))
+                                   (v128.const f32x4 0 -0 -0 -0))
+(assert_return (invoke "f32x4.min" (v128.const f32x4 -0 -0 -0 -0)
+                                   (v128.const f32x4 +0 +0 +0 +0))
+                                   (v128.const f32x4 -0 -0 -0 -0))
+(assert_return (invoke "f32x4.max" (v128.const f32x4 0 0 -0 +0)
+                                   (v128.const f32x4 +0 -0 +0 -0))
+                                   (v128.const f32x4 0 0 0 0))
+(assert_return (invoke "f32x4.max" (v128.const f32x4 -0 -0 -0 -0)
+                                   (v128.const f32x4 +0 +0 +0 +0))
+                                   (v128.const f32x4 +0 +0 +0 +0))
+
+
```

Comment:
I believe these assertions are too strict, as the min/max [meta operations](https://webassembly.github.io/spec/core/exec/numerics.html#op-fmin) applied to a canonical NaN result in a canonical NaN with non-deterministic sign.

Reply:
FYI I don't think there's a practical way to write a test like this right now, since `assert_canonical_nan_fNxM` looks at all the lanes.

Reply:
Ummmm... Maybe we can check each lane with  `f32x4.extract_lane` as a workaround, following example is already used in [simd_f32x4_arith.wast](https://github.com/WAVM/WAVM/blob/master/Test/spec/simd/simd_f32x4_arith.wast#L4303):

```
  (func $f32x4_sqrt_canon (result v128)
    v128.const f32x4 -1.0 nan 4.0 9.0
    f32x4.sqrt)
  (func (export "f32x4_extract_lane_canon_0") (result f32)
    (f32x4.extract_lane 0 (call $f32x4_sqrt_canon)))
  (func (export "f32x4_extract_lane_canon_1") (result f32)
    (f32x4.extract_lane 1 (call $f32x4_sqrt_canon)))
  (func (export "f32x4_extract_lane_canon_2") (result f32)
    (f32x4.extract_lane 2 (call $f32x4_sqrt_canon)))
  (func (export "f32x4_extract_lane_canon_3") (result f32)
    (f32x4.extract_lane 3 (call $f32x4_sqrt_canon)))

(assert_return_canonical_nan (invoke "f32x4_extract_lane_canon_0"))
(assert_return_canonical_nan (invoke "f32x4_extract_lane_canon_1"))
(assert_return (invoke "f32x4_extract_lane_canon_2") (f32.const 2.0))
(assert_return (invoke "f32x4_extract_lane_canon_3") (f32.const 3.0))

```

Reply:
I think that's an ok solution for now, though it's obviously very cumbersome.

It might also be good to use a table + `call_indirect` to ensure that the VM cannot inline the call and combine the vector op with the extract lane op to make a scalar op:
```wasm
(module
  (type $vv_v (func (param v128 v128) (result v128))
  (table funcref (elem $f32x4_min $f32x4_max))
  
  (func $f32x4_min (type $vv_v)
    (f32x4.min (local.get 0) (local.get 1)))
  
  (func $f32x4_max (type $vv_v)
    (f32x4.max (local.get 0) (local.get 1)))
  
  (func (export "call_indirect_vv_v_f32x4_extract_lane_0")
    (param i32 v128 v128) (result f32)
    (f32x4.extract_lane 0
      (call_indirect (type $vv_v) (local.get 0) (local.get 1) (local.get 2))))
  (func (export "call_indirect_vv_v_f32x4_extract_lane_1")
    (param i32 v128 v128) (result f32)
    (f32x4.extract_lane 1
      (call_indirect (type $vv_v) (local.get 0) (local.get 1) (local.get 2))))
  (func (export "call_indirect_vv_v_f32x4_extract_lane_2")
    (param i32 v128 v128) (result f32)
    (f32x4.extract_lane 2
      (call_indirect (type $vv_v) (local.get 0) (local.get 1) (local.get 2))))
  (func (export "call_indirect_vv_v_f32x4_extract_lane_3")
    (param i32 v128 v128) (result f32)
    (f32x4.extract_lane 3
      (call_indirect (type $vv_v) (local.get 0) (local.get 1) (local.get 2))))
)

(assert_return_canonical_nan
  (invoke "call_indirect_vv_v_f32x4_extract_lane_0"
    (i32.const 0) ;; f32x4_min
    (v128.const f32x4 nan 0 0 1)
    (v128.const f32x4 0 nan 1 0)))
(assert_return_canonical_nan
  (invoke "call_indirect_vv_v_f32x4_extract_lane_1"
    (i32.const 0) ;; f32x4_min
    (v128.const f32x4 nan 0 0 1)
    (v128.const f32x4 0 nan 1 0)))
(assert_return
  (invoke "call_indirect_vv_v_f32x4_extract_lane_2"
    (i32.const 0) ;; f32x4_min
    (v128.const f32x4 nan 0 0 1)
    (v128.const f32x4 0 nan 1 0))
  (f32.const 0))
(assert_return
  (invoke "call_indirect_vv_v_f32x4_extract_lane_3"
    (i32.const 0) ;; f32x4_min
    (v128.const f32x4 nan 0 0 1)
    (v128.const f32x4 0 nan 1 0))
  (f32.const 0))
```

Reply:
@AndrewScheidecker, good proposal! But unfortunately, `call_indirect` does not support v128 as a param at present.  Are you going to implement it in WAVM?

Reply:
> But unfortunately, call_indirect does not support v128 as a param at present. Are you going to implement it in WAVM?

It does support `v128` as a parameter, but I messed up the operand order in my example. The `i32` table index should be the last operand of the `call_indirect` instruction, after the call arguments.

Reply:
Ops, I messed up too. Works now!